javascript/diff_match_patch_uncompressed.js

```diff
@@ -1340,11 +1340,19 @@ diff_match_patch.prototype.diff_levenshtein = function(diffs) {
 };
 
 diff_match_patch.prototype.isHighSurrogate = function(c) {
+  if(typeof c !== 'string'){
+    return false
+  }
```

Comment:
this looks confusing from an API point of view. if we're chopping off strings and getting `undefined` would it make more sense to test before sending here?

```js
var leadingOffset = patch.start2 - padding;
var prefix = leadingOffset >= 0 && this.isLowSurrogate(text[leadingOffset])
	? text.substring(leadingOffset - 1, patch.start2)
	: text.substring(leadingOffset, patch.start2)
```

I just feel like we're solving a problem that `isHighSurrogate` doesn't have but leaving no evidence for the need to solve it where the problem appears, in the patch generation.

alternatively we could solve a few related issues if we moved this logic into its own function. works inline too, but once I start seeing more complicated logic…

```js
/**
 * Prepares a contextual prefix for displaying the patch.
 *
 * Note: if prefix would otherwise split the input document
 *       at non-character boundaries, this will attempt to
 *       expand the prefix by a character to complete the
 *       Unicode code point. if it's unable to find a whole
 *       code point it will instead replace the start of the
 *       prefix with U+FFFD, the Unicode replacement character.
 */
function patch_contextPrefix(text, patch, padding) {
	var offset = patch.start2 - padding;

	if (offset < 0) {
		return '';
	}

	var prefix = text.substring(offset, patch.start2);

	// if not splitting a character boundary
	// we don't have to augment our prefix
	if (!this.isLowSurrogate(prefix[0])) {
		return prefix;
	}

	// if the prefix character is the paired surrogate
	// we can expand the prefix by a single character
	if (offset > 0 && this.isHighSurrogate(text[offset - 1])) {
		return text.substring(offset - 1, patch.start2);
	}

	// otherwise give up and replace the lone surrogate
	// with something indicating it's unrepresentable.
	// (this is for display, not for diffing)
	return '\uFFFD' + prefix.slice(1);
}
```

```js
var prefix = this.patch_contextPrefix(text, patch, padding);
```

thoughts?